Create an App Engine app

ivannaranjo opened the pull request:
This PR adds the functionality to create a new App Engine app directly from Visual Studio when required. There are two entrypoints to the new functionality:
* A right click menu in the placeholder shown when no app is found in the Cloud Explorer.
* During a deployment to App Engine flexible, when the extension detects that there's no app.

This PR also merges the functionality of the `Polling<T>` class into the `OperationUtils` so we have a single way of polling for operations moving forward.

Fixes #634 Fixed #508

ILMTitan commented on GoogleCloudExtension/GoogleCloudExtension.DataSources/OperationUtils.cs:
```diff
@@ -35,29 +36,52 @@ public static class OperationUtils
+                    // Check the timeout.
+                    if (watch != null)
+                    {
+                        if (watch.Elapsed >= timeout)
```
These nested ifs could just be `if(watch?.Elapsed >= timeout)`. `null` compares false

ivannaranjo replied:
I like to be explicit. In this case the if is checking if there was a timeout to begin with.

ivannaranjo replied:
The world of lifted operations is amazing :).

ILMTitan commented on GoogleCloudExtension/GoogleCloudExtension.DataSources/OperationUtils.cs:
```diff
@@ -35,29 +36,52 @@ public static class OperationUtils
         /// <param name="getErrorData">A function that will extract the error message from the operation, if in error.</param>
+        /// <param name="token">The cancellation token to stop polling for the state of the operation. This does not cancel the operation.</param>
         /// <param name="delay">The delay to use in between refreshes.</param>
+        /// <param name="timeout">The timeout for the operation. If null there's no timeout.</param>
```
Missing documentation for `TErrorData` and `getErrorMessage`.

ivannaranjo replied:
Done.

ILMTitan commented on GoogleCloudExtension/GoogleCloudExtension/AppEngineManagement/AppEngineManagementViewModel.cs:
```diff
@@ -0,0 +1,76 @@
+
+namespace GoogleCloudExtension.AppEngineManagement
+{
+    public class AppEngineManagementViewModel : ViewModelBase
```
Could we get some unit tests for this class? You will only need to inject and mock the GaeDataSource.

ILMTitan commented on GoogleCloudExtension/GoogleCloudExtension/AppEngineManagement/AppEngineManagementWindowContent.xaml:
```diff
@@ -0,0 +1,49 @@
+            <Label Target="{Binding ElementName=_locations}"
+                   Content="{x:Static ext:Resources.AppEngineManagementRegionsLabel}"
+                   Style="{StaticResource CommonLabelStyle}" />
+            <ComboBox x:Name="_locations"
```
It feels like this needs a loading placeholder. Not everyone has a fast connection.

ivannaranjo replied:
Will think on a more general model. We don't really have that for all dialogs and would like to be consistent. Perhaps a dialog that shows up after a few seconds?

ILMTitan commented on GoogleCloudExtension/GoogleCloudExtension/AppEngineManagement/AppEngineManagementViewModel.cs:
```diff
@@ -0,0 +1,76 @@
+            _owner = owner;
+
+            Locations = new AsyncProperty<IEnumerable<string>>(ListAllLocationsAsync());
+            ActionCommand = new ProtectedCommand(OnActionCommand);
```
Consider disabling this command until the load is complete.

ivannaranjo replied:
The entire dialog is disabled until the load is completed.

ILMTitan commented on GoogleCloudExtension/GoogleCloudExtension/AppEngineManagement/AppEngineManagementViewModel.cs:
```diff
@@ -0,0 +1,76 @@
+
+            Locations = new AsyncProperty<IEnumerable<string>>(ListAllLocationsAsync());
+            ActionCommand = new ProtectedCommand(OnActionCommand);
+            Message = string.Format(Resources.AppEngineManagementAppCreationMessage, projectId);
```
Message isn't really data, and shouldn't be on the ViewModel. Replace it with a bare `ProjectId` property.
See my comment on the view on how to use it.

ivannaranjo replied:
View model does any transformation required for the view, not limited to only data.

ILMTitan commented on GoogleCloudExtension/GoogleCloudExtension/AppEngineManagement/AppEngineManagementWindowContent.xaml:
```diff
@@ -0,0 +1,49 @@
+
+        <StackPanel>
+            <TextBlock Style="{StaticResource CommonTextStyle}"
+                       Text="{Binding Message, Mode=OneTime}"
```
Replace this with `Text="{Binding ProjectId, StringFormat={x:Static ext:Resources.AppEngineManagementAppCreationMessage}, Mode=OneTime}"`

The resource string is part of the presentation, and should be here on the view.

ivannaranjo replied:
No, view models can do any transformation required for view. The view does not have to do this.

ivannaranjo replied:
Done.

ILMTitan commented on GoogleCloudExtension/GoogleCloudExtension/CloudExplorerSources/Gae/GaeSourceRootViewModel.cs:
```diff
@@ -181,8 +176,21 @@ private GaeDataSource CreateDataSource()
                 GaeApplication = await _dataSource.Value.GetApplicationAsync();
                 if (GaeApplication == null)
                 {
+                    TreeLeaf noAppEngineAppPlaceholder = new TreeLeaf
```
This may be `var`, according to the style guide.

ILMTitan commented on GoogleCloudExtension/GoogleCloudExtension/AppEngineManagement/AppEngineManagementViewModel.cs:
```diff
@@ -0,0 +1,76 @@
+        private async Task<IEnumerable<string>> ListAllLocationsAsync()
+        {
+            var source = new GaeDataSource(
+                CredentialsStore.Default.CurrentProjectId,
```
Is it ok that this project may be different from the input project?

ivannaranjo replied:
It is getting whatever is the current project so it will always match the project selected in the publish dialog. You do bring a good point, if the user changes the project in the publish dialog then we need to repeat all of the checks to ensure that the new project is also ready to deployment. Need to rethink this model.

ivannaranjo commented:
PTAL, I have reworked the way the dialogs indicate that APIs are enabled, added tests for all view models, and reworked the way the app is created.

See for example the App Engine publishing step:
<img width="515" alt="screen shot 2017-11-16 at 3 25 25 pm" src="https://user-images.githubusercontent.com/9807532/32899287-748fd544-cae2-11e7-9d78-09310f458796.png">

Also see the GKE step for the "needs API enabled" state:
<img width="519" alt="screen shot 2017-11-16 at 3 26 41 pm" src="https://user-images.githubusercontent.com/9807532/32899328-9abe0f4c-cae2-11e7-86bd-de68bcb3764b.png">

I have also refactored the GKE, GCE and GAE data sources to make them friendly to dependency injection and use that for tests in the view models for the various steps.

ILMTitan commented on GoogleCloudExtension/GoogleCloudExtension/PublishDialogSteps/GkeStep/GkeStepContent.xaml:
```diff
@@ -34,111 +36,136 @@
                            Style="{StaticResource CommonHyperlinkStyle}">
                     <TextBlock Text="{x:Static ext:Resources.GkePublishNewClusterCaption}" />
                 </Hyperlink>
             </TextBlock>
```
These lines also need to be indented.

ivannaranjo replied:
Done.

ILMTitan commented on GoogleCloudExtension/GoogleCloudExtension/PublishDialogSteps/GkeStep/GkeStepContent.xaml:
```diff
@@ -34,111 +36,136 @@
                            Style="{StaticResource CommonHyperlinkStyle}">
                     <TextBlock Text="{x:Static ext:Resources.GkePublishRefreshClustersCaption}" />
                 </Hyperlink>
             </TextBlock>
```
These lines also need to be indented.

ivannaranjo replied:
Done.

ILMTitan commented on GoogleCloudExtension/GoogleCloudExtension/PublishDialogSteps/GkeStep/GkeStepContent.xaml:
```diff
@@ -34,111 +36,136 @@
+                       Command="{Binding EnableApiCommand}">
+                <TextBlock Text="{x:Static ext:Resources.PublishDialogEnableApiCaption}" />
+            </Hyperlink>
+        </TextBlock>
```
The 5 lines above need to be indented.

ivannaranjo replied:
Done.

ILMTitan commented on GoogleCloudExtension/GoogleCloudExtension/Resources.resx:
```diff
@@ -2836,7 +2836,7 @@
   </data>
   <data name="CloudExplorerGaeNoAppFoundCaption" xml:space="preserve">
-    <value>No App Engine app found.</value>
+    <value>No app engine app found.</value>
```
I think the capitalization here is correct. App Engine is a proper name, like Cloud SQL or Google Cloud Platform.

ivannaranjo replied:
Agreed, not sure what happened there.

ILMTitan commented on GoogleCloudExtension/GoogleCloudExtension/Resources.resx:
```diff
@@ -2903,4 +2887,56 @@
+    <value>Available _regions:</value>
+    <comment>Label to be used for the regions drop box.</comment>
+  </data>
+  <data name="AppEngineManagementWindowCaption" xml:space="preserve">
```
Can we call this "Title" rather than "Caption"? The [casing guidelines](https://msdn.microsoft.com/en-us/library/mt162316.aspx#BKMK_TextStyle) use "Title".

ivannaranjo replied:
Renamed to title to be more consistent with other windows as well. Not sure about the casing rules now?

ILMTitan commented on GoogleCloudExtension/GoogleCloudExtension/Resources.resx:
```diff
@@ -2903,4 +2887,56 @@
+    <comment>Caption for the select region dialog.</comment>
+  </data>
+  <data name="CloudExplorerGaeSetAppRegionMenuHeader" xml:space="preserve">
+    <value>Set app region</value>
```
Menu Items should be Title Case.

ivannaranjo replied:
Yep, and needed to add the ... since it opens a dialog.

ILMTitan commented on GoogleCloudExtension/GoogleCloudExtension/Resources.resx:
```diff
@@ -2903,4 +2887,56 @@
+    <value>Set app region</value>
+    <comment>Menu header for the menu shown to the user to set the app's region.</comment>
+  </data>
+  <data name="GaeUtilsSetAppEngineRegionProgressCaption" xml:space="preserve">
```
Caption => Title

ivannaranjo replied:
Done.

ILMTitan commented on GoogleCloudExtension/GoogleCloudExtension/Resources.resx:
```diff
@@ -2903,4 +2887,56 @@
+  </data>
+  <data name="GaeUtilsSetAppEngineRegionProgressMessage" xml:space="preserve">
+    <value>Setting App Engine region.</value>
+    <comment>The message for the progress dialog shwon while setting the app engine region.</comment>
```
shwon => shown

ILMTitan commented on GoogleCloudExtension/GoogleCloudExtension/Resources.resx:
```diff
@@ -2903,4 +2887,56 @@
+    <value>Loading regions...</value>
+    <comment>String used as a placeholder while loading the list of regions in the dialog.</comment>
+  </data>
+  <data name="PublishDialogEnableApiCaption" xml:space="preserve">
```
Caption => Title

ivannaranjo replied:
This is a hyperlink, not a menu item.

ILMTitan commented on GoogleCloudExtension/GoogleCloudExtension/Resources.resx:
```diff
@@ -2903,4 +2887,56 @@
+    <comment>String used as a placeholder while loading the list of regions in the dialog.</comment>
+  </data>
+  <data name="PublishDialogEnableApiCaption" xml:space="preserve">
+    <value>Enable the services</value>
```
Title Case

ivannaranjo replied:
This is hyperlink not a menu tiem.

ILMTitan commented on GoogleCloudExtension/GoogleCloudExtension/Utils/GaeUtils.cs:
```diff
@@ -47,5 +53,36 @@ public static string GetAppUrl(string hostname, string serviceId = null, string
+            string selectedLocation = AppEngineManagementWindow.PromptUser(projectId);
+            if (selectedLocation == null)
+            {
+                Debug.WriteLine("The user cancelled creating a new app");
```
End this sentence with a "."?

ivannaranjo replied:
It's an internal debug message, but ok :)

ILMTitan commented on GoogleCloudExtension/GoogleCloudExtension/Theming/ICloseable.cs:
```diff
@@ -0,0 +1,29 @@
+    /// Interface to be implemented by those Windows that can be closed. Allows for mocking
+    /// of the window.
+    /// </summary>
+    public interface ICloseable
```
I tend to prefer sending `Action closeAction = window.Close` as a view model constructor parameter, but in the end it doesn't matter much either way.

ivannaranjo replied:
This is a tad more specific, we only allow things that are closeable to be passed in rather than an action that could perform other actions. We should converge into a single pattern though.

ILMTitan commented on GoogleCloudExtension/GoogleCloudExtension/Resources.resx:
```diff
@@ -2903,4 +2887,56 @@
+    <comment>The message shown in the steps while the project is being loaded.</comment>
+  </data>
+  <data name="PublishDialogProjectNeedsApiMessage" xml:space="preserve">
+    <value>The selected GCP project needs services to be enabled before you can deploy.</value>
```
Would this be better taking the actual project id as a parameter?

ivannaranjo replied:
It is shown right under the textbox with the project id, I think it would be redundandt.

ILMTitan commented on GoogleCloudExtension/GoogleCloudExtensionUnitTests/AppEngineManagement/AppEngineManagementViewModelTests.cs:
```diff
@@ -0,0 +1,96 @@
+
+            _testedViewModel.SelectedLocation = s_mockFlexLocations.First();
+
+            Assert.IsTrue(_testedViewModel.ActionCommand.CanExecute(null));
```
Stop this test here. For the code below, create a new test, using the above setup and action as the setup.

You probably want to use this name for the next test, and rename this test "SelectionTest".

ivannaranjo replied:
What I want to test is the result, why would I split this test into two? That seems like it would make things overly complicated.

ivannaranjo replied:
Done.

ILMTitan commented on GoogleCloudExtension/GoogleCloudExtensionUnitTests/AppEngineManagement/AppEngineManagementViewModelTests.cs:
```diff
@@ -0,0 +1,96 @@
+
+            _mockedWindow.Verify(x => x.Close(), Times.Exactly(1), "Failed to close the window on action.");
+        }
+    }
```
Perhaps add a test with `_flexLoactionsSource.SetException(new DataSourceException())`?

ivannaranjo replied:
The view model doesn't do much with that, other than show an error message. (now).

ILMTitan commented on GoogleCloudExtension/GoogleCloudExtensionUnitTests/PublishDialogSteps/FlexStep/FlexStepViewModelTests.cs:
```diff
@@ -13,26 +13,138 @@
+
+            _areServicesEnabledTaskSource = new TaskCompletionSource<bool>();
+            _appTaskSource = new TaskCompletionSource<Google.Apis.Appengine.v1.Data.Application>();
+            
```
remove trailing whitespace.

ILMTitan commented on GoogleCloudExtension/GoogleCloudExtensionUnitTests/PublishDialogSteps/FlexStep/FlexStepViewModelTests.cs:
```diff
@@ -13,26 +13,138 @@
+            Assert.IsFalse(_objectUnderTest.NeedsAppCreated);
+            Assert.IsFalse(_objectUnderTest.GeneralError);
+
+            // Mock a positive validation.
```
Start a new test.

ivannaranjo replied:
I don't understand this comment. Why would I start a new test?

ILMTitan commented on GoogleCloudExtension/GoogleCloudExtensionUnitTests/PublishDialogSteps/FlexStep/FlexStepViewModelTests.cs:
```diff
@@ -13,26 +13,138 @@
+
+            // Check that the expected methods were called.
+            _mockedGaeDataSource.Verify(x => x.GetApplicationAsync(), Times.AtLeastOnce);
+            _mockedApiManager.Verify(x => x.AreServicesEnabledAsync(It.IsAny<IList<string>>()), Times.AtLeastOnce);
```
These calls are an implementation detail, not part of the API. The calls should change the public state of the object under test, and the public state should be the thing we test.

ivannaranjo replied:
But I do want to make sure that this calls are made, I want to make sure that the view model tried to fetch the application and checked that the services are enabled. They might be implementation detail, but I would consider it a failure if the view model didn't call `AreServicesEnabledAsync()`≥

ILMTitan replied:
The idea is to test that call by seeing how different return values affect the public state of the view model.

ILMTitan commented on GoogleCloudExtension/GoogleCloudExtensionUnitTests/PublishDialogSteps/FlexStep/FlexStepViewModelTests.cs:
```diff
@@ -13,26 +13,138 @@
+        public async Task TestNeedsApiState()
+        {
+            _objectUnderTest.OnPushedToDialog(_mockedPublishDialog.Object);
+            Assert.IsNotNull(_objectUnderTest.LoadingProjectTask);
```
You don't need to repeat this assertion. It is already known to be true from another test.

ivannaranjo replied:
Does it hurt?

ILMTitan commented on GoogleCloudExtension/GoogleCloudExtensionUnitTests/PublishDialogSteps/GceStep/GceStepViewModelTests.cs:
```diff
@@ -0,0 +1,131 @@
+            // Check the state before the validation.
+            Assert.IsTrue(_objectUnderTest.LoadingProject);
+            Assert.IsFalse(_objectUnderTest.NeedsApiEnabled);
+
```
Start a new test here.

ivannaranjo replied:
But why?

ILMTitan commented on GoogleCloudExtension/GoogleCloudExtensionUnitTests/AppEngineManagement/AppEngineManagementViewModelTests.cs:
```diff
@@ -0,0 +1,96 @@
+
+            Assert.AreEqual(s_mockFlexLocations.First(), _testedViewModel.Result);
+
+            _mockedWindow.Verify(x => x.Close(), Times.Exactly(1), "Failed to close the window on action.");
```
Times.Exactly(1) === Times.Once()

ivannaranjo replied:
Done.

ILMTitan commented on GoogleCloudExtension/GoogleCloudExtensionUnitTests/PublishDialogSteps/GceStep/GceStepViewModelTests.cs:
```diff
@@ -0,0 +1,131 @@
+            // Verify the expected methods were called.
+            _mockedApiManager.Verify(x => x.AreServicesEnabledAsync(It.IsAny<IList<string>>()), Times.AtLeastOnce);
+            _mockedDataSource.Verify(x => x.GetInstanceListAsync(), Times.AtLeastOnce);
+            _mockedPublishDialog.Verify(x => x.TrackTask(It.IsAny<Task>()), Times.AtLeastOnce);
```
Implementation details. Remove these.

ivannaranjo replied:
As I explained before, I want to know that these methods were called. I want to make sure that the VM did in fact reach into the mock for the data.

ILMTitan commented on GoogleCloudExtension/GoogleCloudExtensionUnitTests/PublishDialogSteps/GceStep/GceStepViewModelTests.cs:
```diff
@@ -0,0 +1,131 @@
+
+            // Check the state before the validation.
+            Assert.IsTrue(_objectUnderTest.LoadingProject);
+            Assert.IsFalse(_objectUnderTest.NeedsApiEnabled);
```
You don't need to assert preconditions. There should be a test already making these assertions.

ILMTitan commented on GoogleCloudExtension/GoogleCloudExtensionUnitTests/PublishDialogSteps/GkeStep/GkeStepViewModelTests.cs:
```diff
@@ -0,0 +1,138 @@
+            // Check the state before the validation.
+            Assert.IsTrue(_objectUnderTest.LoadingProject);
+            Assert.IsFalse(_objectUnderTest.NeedsApiEnabled);
+
```
Start new test here.

ILMTitan commented on GoogleCloudExtension/GoogleCloudExtensionUnitTests/PublishDialogSteps/GkeStep/GkeStepViewModelTests.cs:
```diff
@@ -0,0 +1,138 @@
+            // Verify the expected calls.
+            _mockedApiManager.Verify(x => x.AreServicesEnabledAsync(It.IsAny<IList<string>>()), Times.AtLeastOnce);
+            _mockedDataSource.Verify(x => x.GetClusterListAsync(), Times.AtLeastOnce);
+            _mockedPublishDialog.Verify(x => x.TrackTask(It.IsAny<Task>()), Times.AtLeastOnce);
```
Remove implementation details.

Deren-Liao commented on GoogleCloudExtension/GoogleCloudExtension/PublishDialogSteps/GceStep/GceStepViewModel.cs:
```diff
@@ -129,18 +132,36 @@ public bool LaunchRemoteDebugger
+            EnableApiCommand = new ProtectedCommand(OnEnableApiCommand);
+        }
+
+        private async void OnEnableApiCommand()
```
unhandled exception inside async void would abruptly terminate Visual Studio.
The way I handle it is to 
EnableApiCommand = new ProtectedAsyncCommand(OnEnableApiCommand);
and change OnEnableApiCommand signature to    async Task

ivannaranjo replied:
That is a good point, replaced all async commands.

ivannaranjo commented:
Thank you for the exhaustive reviews, I will address the comments and get back to you.

ivannaranjo commented on GoogleCloudExtension/GoogleCloudExtension.DataSources/IGaeDataSource.cs:
```diff
@@ -0,0 +1,117 @@
+        /// </summary>
+        /// <param name="serviceId">The id of the service</param>
+        /// <returns>The list of GAE versions.</returns>
+        Task<IList<Google.Apis.Appengine.v1.Data.Version>> GetVersionListAsync(string serviceId);
```
Indeed.

ivannaranjo replied:
Done.

ivannaranjo commented on GoogleCloudExtension/GoogleCloudExtension.DataSources/IGaeDataSource.cs:
```diff
@@ -0,0 +1,117 @@
+        /// <param name="serviceId">The id of the service</param>
+        /// <param name="versionId">The id of the version</param>
+        /// <returns>The GAE version.</returns>
+        Task<Google.Apis.Appengine.v1.Data.Version> GetVersionAsync(string serviceId, string versionId);
```
Same.

ivannaranjo commented on GoogleCloudExtension/GoogleCloudExtension.DataSources/OperationUtils.cs:
```diff
@@ -31,33 +32,57 @@ public static class OperationUtils
-            TimeSpan? delay = null)
+            Func<TOperation, TErrorData> getErrorData,
+            Func<TErrorData, string> getErrorMessage,
+            CancellationToken token = default(CancellationToken),
```
True. Done.

ivannaranjo commented on GoogleCloudExtension/GoogleCloudExtension.DataSources/OperationUtils.cs:
```diff
@@ -31,33 +32,57 @@ public static class OperationUtils
+            Stopwatch watch = null;
+            if (timeout != null)
+            {
+                watch = Stopwatch.StartNew();
```
I one-upped your proposal. I have removed the `timeout` parameter. The caller can do this if they want. This will simplify the generic.

This call came from combining two previously different implementations of this algorithm.

ivannaranjo commented on GoogleCloudExtension/GoogleCloudExtension/AppEngineManagement/AppEngineManagementViewModel.cs:
```diff
@@ -0,0 +1,115 @@
+        /// <summary>
+        /// The placeholder for the list of regions being loaded.
+        /// </summary>
+        internal static readonly IEnumerable<string> s_loadingPlaceholder = new string[] 
```
This variable is internal so it can be accessed by tests.

Static variables do use the `s_` prefix.

ivannaranjo replied:
I added a comment to indicate this.

ivannaranjo commented on GoogleCloudExtension/GoogleCloudExtension/AppEngineManagement/AppEngineManagementViewModel.cs:
```diff
@@ -0,0 +1,115 @@
+        /// <summary>
+        /// The placeholder for the list of regions being loaded.
+        /// </summary>
+        internal static readonly IEnumerable<string> s_loadingPlaceholder = new string[] 
```
Ack

ivannaranjo commented on GoogleCloudExtension/GoogleCloudExtension/AppEngineManagement/AppEngineManagementWindowContent.xaml:
```diff
@@ -0,0 +1,57 @@
+             xmlns:local="clr-namespace:GoogleCloudExtension.AppEngineManagement"
+             xmlns:theming="clr-namespace:GoogleCloudExtension.Theming"
+             xmlns:ext="clr-namespace:GoogleCloudExtension"
+             xmlns:utils="clr-namespace:GoogleCloudExtension.Utils;assembly=GoogleCloudExtension.Utils"
```
Done.

ivannaranjo commented on GoogleCloudExtension/GoogleCloudExtension/CloudExplorerSources/CloudSQL/InstanceViewModel.cs:
```diff
@@ -109,9 +109,7 @@ private void OnPropertiesCommand()
-                // Poll until the update to completes.
-                var operation = await _owner.DataSource.Value.UpdateInstanceAsync(Instance);
-                await _owner.DataSource.Value.AwaitOperationAsync(operation);
+                await _owner.DataSource.Value.UpdateInstanceAsync(Instance);
```
Old code, not modified.

ivannaranjo commented on GoogleCloudExtension/GoogleCloudExtension/CloudExplorerSources/Gae/ServiceViewModel.cs:
```diff
@@ -333,8 +333,7 @@ private void OnOpenService()
             {
-                var operation = await _owner.DataSource.UpdateServiceTrafficSplitAsync(split, Service.Id);
-                await _owner.DataSource.AwaitOperationAsync(operation);
+                await _owner.DataSource.UpdateServiceTrafficSplitAsync(split, Service.Id);
```
Same.

ivannaranjo commented on GoogleCloudExtension/GoogleCloudExtension/PublishDialog/PublishDialogStepBase.cs:
```diff
@@ -80,13 +87,62 @@ protected set
+            protected set
+            {
+                SetValueAndRaise(ref _needsApiEnabled, value);
+                RaisePropertyChanged(nameof(ShowInputControls));
```
This logic is something that every step needs, which is why is moved up in the class hierarchy.

I like your suggestion to share the logic. Done.

ivannaranjo commented on GoogleCloudExtension/GoogleCloudExtension/PublishDialog/PublishDialogStepBase.cs:
```diff
@@ -80,13 +87,62 @@ protected set
+        /// Whether the project is loaded, which include validating that the project is correctly
+        /// setup for deployment and loading the necessary data to display to the user.
+        /// </summary>
+        public bool LoadingProject
```
Loading project is not a value, is a state flag to denote that we're actually loading the project. As such making it an AsyncProperty would be awkward in my opinion. That's why I did it this way.

ILMTitan replied:
The non-generic `AsyncProperty` does not have a value, it just tracks the result of a non-generic `Task`. It is literally just a set of state flags.

ivannaranjo commented on GoogleCloudExtension/GoogleCloudExtension/PublishDialog/PublishDialogStepBase.cs:
```diff
@@ -80,13 +87,62 @@ protected set
+        /// <summary>
+        /// The task that tracks the project loading process.
+        /// </summary>
+        internal Task LoadingProjectTask { get; set; }
```
Not quite, as it is not a value but a state flag.

ivannaranjo commented on GoogleCloudExtension/GoogleCloudExtension/PublishDialogSteps/GceStep/GceStepViewModel.cs:
```diff
@@ -40,13 +41,15 @@ namespace GoogleCloudExtension.PublishDialogSteps.GceStep
 
         private readonly GceStepContent _content;
+        private readonly IGceDataSource _dataSource;
+        private readonly IApiManager _apiManager;
```
Done.

ivannaranjo commented:
PTAL, I think that the only contentious part is how the tests are written.

ivannaranjo commented:
PTAL, I have simplified the tests, make them more predictable and simplified the error conditions on the VM as a result. I actually liked this process, made it pretty apparent that the code needed better handling of error conditions.
